tests/unit/test_unit_smhi.py

```diff
@@ -0,0 +1,249 @@
+        client = SMHI()
+        data = client.get_data_by_city(parameter, city, distance)
+        assert data == "test"
+
+    distanceresponse = MagicMock()
+    distanceresponse.km = 0
+
+    @pytest.mark.parametrize(
+        "latitude, longitude,dist",
+        [
+            (
+                59,
+                17,
+                0,
+            ),
+            (
+                59.4,
+                17,
+                30,
+            ),
+        ],
+    )
+    @patch("geopy.distance.distance", return_value=distanceresponse)
+    def test_find_stations_from_gps(self, mock_distance, latitude, longitude, dist):
+        """Unit test for SMHI find_stations_from_gps method.
+
+        Args:
+        """
+        station1 = MagicMock()
+        station1.id = 1
+        station1.name = "Akalla"
+        station1.latitude = 59.5
+        station1.longitude = 17.8
+        stations = MagicMock()
+        stations.station = [station1]
+
+        client = SMHI()
+        nearby_town = client._find_stations_from_gps(
+            stations, latitude, longitude, dist
+        )
+        assert nearby_town[0][0] == 1
+
+    @pytest.mark.parametrize(
+        "parameter, city, distance", [(8, "Bengtsfors", None), (8, "Bengtsfors", 50)]
+    )
+    @patch("smhi.metobs.Stations.__new__")
+    @patch("smhi.metobs.Periods.__new__")
+    @patch("smhi.metobs.Data.__new__")
+    @patch("geopy.geocoders.Nominatim.__new__")
+    @patch("smhi.smhi.SMHI._find_stations_from_gps")
+    def test_find_stations_by_city(
+        self,
+        mock_find_from_gps,
+        mock_nominatim,
+        mock_data_data,
+        mock_period_data,
+        mock_station_data,
+        parameter,
+        city,
+        distance,
+    ):
+        """Unit test for SMHI _find_stations_by_city method.
+
+        Args:
+        """
+        client = SMHI()
+        _ = client._find_stations_by_city(parameter, city, distance)
+        mock_nominatim.assert_called_once()
+
+    @pytest.mark.parametrize("distance", [(0), (50)])
+    @patch("smhi.metobs.Stations.__new__")
+    @patch("smhi.metobs.Periods.__new__")
+    @patch("smhi.metobs.Data.__new__")
+    @patch("smhi.smhi.SMHI._iterate_over_time")
+    @patch("smhi.smhi.SMHI._find_missing_data")
+    @patch("smhi.smhi.SMHI._find_stations_from_gps")
+    def test_interpolate(
+        self,
+        mock_find_from_gps,
+        mock_find_missing_data,
+        mock_iterate_over_time,
+        mock_data_data,
+        mock_period_data,
+        mock_station_data,
+        distance,
+    ):
+        """Unit test for SMHI _interpolate method.
+
+        Args:
+        """
+        client = SMHI()
+        data = client._interpolate(
+            distance, mock_station_data, mock_period_data, mock_data_data
+        )
+        if distance <= 0:
+            assert data == mock_data_data
+        else:
+            mock_find_from_gps.assert_called_once()
+
+    def test_iterate_over_time(
```

Comment:
Maybe we could use `pytest.fixtures` to set-up the test data which would make this test much smaller.

Reply:
Maybe.

Reply:
A fixture is just this

```python
@pytest.fixture
def setup_test_x():
    df = pd.DataFrame(
            {
                "date": [
                    "2024-04-21 10:00",
                    "2024-04-21 11:00",
                    "2024-04-21 12:00",
                    "2024-04-22 12:00",
                ],
                "Temperatur": [1, 1, 1, 12],
            }
        )
        df["date"] = pd.to_datetime(df["date"])
        df = df.set_index("date")

        nearby_df = pd.DataFrame(
            {"date": ["2024-04-22 10:00", "2024-04-22 11:00"], "Temperatur": [8, 9]}
        )
        nearby_df["date"] = pd.to_datetime(nearby_df["date"])
        nearby_df = nearby_df.set_index("date")

        missing_df = df[
            df.index.to_series().diff() > df.index.to_series().diff().median()
        ]

    return df, nearby_df, missing_df
```

And then in your test you include it as one of the parameters and unpack it

```python
def test_iterate_over_time(self, setup_test_x)
    df, nearby_df, missing_df = setup_test_x
```

It's just a suggestion to keep the test function clean and simple and move data preparation stuff out of it.

Reply:
I will give this a go, too.